boards: arm: add support for NUCLEO-L031K6 board

sdaglish opened the pull request:
Support for the NUCLEO-L031K6 development board with STM32L031K6 SoC.

Although the SoC only contains 32k flash and 8K RAM, it has been tested
on a number of samples and has worked as expected.

Signed-off-by: Steven Daglish <s.c.daglish@gmail.com>

erwango commented on boards/arm/nucleo_l031k6/Kconfig.board:
```diff
@@ -0,0 +1,8 @@
+# STM32 Nucleo-32 development board with STM32L013K6 MCU configuration
```
```suggestion
# STM32 Nucleo-32 development board with STM32L031K6 MCU configuration
```

sdaglish replied:
Typo fixed

erwango commented on samples/sensor/mcp9808/boards/nucleo_l031k6.overlay:
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright (c) 2019, Linaro Ltd.
```
```suggestion
 * Copyright (c) 2020 Steven Daglish <s.c.daglish@gmail.com>
```

erwango reviewed (changes requested):
Aside from minor nitpicks this looks good to me, but one request:
Please split into 3 commits:
- First commit to add SoC support (soc/, dts/)
- Second commit to add board support (boards/)
- Third commit for the sample (samples/)

sdaglish committed:
boards: arm: add support for nucleo-l031k6 board

Support for the NUCLEO-L031K6 development board with STM32L031K6 SoC.

Although the SoC only contains 32k flash and 8K RAM, it has been tested
on a number of samples and has worked as expected.

Signed-off-by: Steven Daglish <s.c.daglish@gmail.com>

sdaglish commented:
> Aside from minor nitpicks this looks good to me, but one request:
> Please split into 3 commits:
> - First commit to add SoC support (soc/, dts/)
> - Second commit to add board support (boards/)
> - Third commit for the sample (samples/)

Split into 3 commits as requested.

MaureenHelm commented on boards/arm/nucleo_l031k6/doc/index.rst:
```diff
@@ -0,0 +1,169 @@
+
+  - Arduino* Nano V3 connectivity
+
+- ARM* mbed*
```
Why is this in the list of hardware components?

erwango replied:
@sdaglish can you fix this ?

sdaglish replied:
I will look into fixing this issue, but I still think it would be better if all the STM32L0 boards had the same documentation. Therefore, I'll create an issue related to this so that all other boards get changed consistently.

sdaglish replied:
Reference to `Arm* mbed*`.
This is now inline with ST Nucleo L432KC documentation for Hardware.

MaureenHelm commented on boards/arm/nucleo_l031k6/doc/index.rst:
```diff
@@ -0,0 +1,169 @@
+  - Mass storage
+  - Debug port
+
+- Support of wide choice of Integrated Development Environments (IDEs) including:
+
+  - IAR
+  - ARM Keil
+  - GCC-based IDEs
```
Why is this in the list of hardware components?

sdaglish replied:
Reference to IDEs.
This is now inline with ST Nucleo L432KC documentation for Hardware.

erwango commented on boards/arm/nucleo_l031k6/pinmux.c:
```diff
@@ -0,0 +1,34 @@
+/*
```
This file should be removed after SPI pinctrl has been converted to dts

erwango commented on boards/arm/nucleo_l031k6/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_PINMUX)
+zephyr_library()
+zephyr_library_sources(pinmux.c)
+zephyr_library_include_directories(${ZEPHYR_BASE}/drivers)
+endif()
```
This file could be removed once pinmux.c file is removed

erwango reviewed (changes requested):
Few more changes required after complete transition to dts pinctrl

erwango commented on boards/arm/nucleo_l031k6/pinmux.c:
```diff
@@ -0,0 +1,34 @@
+	{STM32_PIN_PA5, STM32L0_PINMUX_FUNC_PA5_SPI1_SCK},
+	{STM32_PIN_PA6, STM32L0_PINMUX_FUNC_PA6_SPI1_MISO},
+	{STM32_PIN_PA7, STM32L0_PINMUX_FUNC_PA7_SPI1_MOSI},
+#endif
```
Please move this to device tree and purely remove the file.

erwango reviewed (changes requested):
Some comments are still pending.
Also please take into account comments from @MaureenHelm

sdaglish reviewed (commented):
@erwango 

> Some comments are still pending.

> This file could be removed once pinmux.c file is removed

I believe that pinmux.c has been completely removed in the latest RP update, and all SPI information is now in device tree.

sdaglish committed:
soc: arm: stm32l0: Add STM32L031 series

These MCUs have 32Kbytes of Flash and 8Kbytes of RAM. They are still
able to run a number of samples.

Signed-off-by: Steven Daglish <s.c.daglish@gmail.com>

sdaglish committed:
sample: sensor: mcp9808: add nucleo-l031k6 overlay to samples

The Nucleo-L031K6 development board has been tested with the MCP9808
sample code and works as expected.

Signed-off-by: Steven Daglish <s.c.daglish@gmail.com>

erwango commented on dts/arm/st/l0/stm32l031X6.dtsi:
```diff
@@ -0,0 +1,22 @@
+ */
+
+#include <mem.h>
+#include <st/l0/stm32l053.dtsi>
```
Sorry for not catching this earlier but this is not correct. This should be <st/l0/stm32l031.dtsi>.
So you need to add st/l0/stm32l031.dtsi.
It could be done by including st/l0/stm32l011.dtsi and adding larger EEPROM.

sdaglish replied:
This should now be fixed.

erwango reviewed (changes requested):
Remains one issue that I initially oversaw and the pinmux change to be completed

erwango commented:
> I believe that pinmux.c has been completely removed in the latest RP update, and all SPI information is now in device tree.

This might be true in you local branch but this is not what I'm seeing here.
Btw, this will fix the CI issue (the #define used in these files are now obsolete and their use generates warnings that trigger CI error)

sdaglish commented:
> > I believe that pinmux.c has been completely removed in the latest RP update, and all SPI information is now in device tree.
> 
> This might be true in you local branch but this is not what I'm seeing here.
> Btw, this will fix the CI issue (the #define used in these files are now obsolete and their use generates warnings that trigger CI error)

This should now be fixed on the PR. PR is now showing no pinmux.c file.

erwango reviewed (changes requested):
Please squash the 4rth commit into the second one.
We're merging fast forward, which means that all your commits will be merged on top of master. It is then expected that all the commits are "first time right" and the is no "fixup commits" in the PR.

sdaglish commented:
> Please squash the 4rth commit into the second one.
> We're merging fast forward, which means that all your commits will be merged on top of master. It is then expected that all the commits are "first time right" and the is no "fixup commits" in the PR.

2nd and 4th commits have now been squashed into a simple commit.
